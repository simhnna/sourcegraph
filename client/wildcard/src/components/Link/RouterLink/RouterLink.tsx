import * as React from 'react'
import { __RouterContext } from 'react-router'
import { Link } from 'react-router-dom'

import { AnchorLink, LinkProps } from '../AnchorLink'

function useInRouterContext(): boolean {
    return Boolean(React.useContext(__RouterContext))
}

export const RouterLink: React.FunctionComponent<LinkProps> = React.forwardRef(
    ({ to, children, ...rest }: LinkProps, reference) => {
        const isInRouter = useInRouterContext()

        if (process.env.NODE_ENV === 'development' && !isInRouter) {
            throw new Error('Please use the `AnchorLink` component outside of `react-router`')
        }

        return (
            <AnchorLink to={to} as={Link} {...rest} ref={reference}>
                {children}
            </AnchorLink>
        )
    }
)
