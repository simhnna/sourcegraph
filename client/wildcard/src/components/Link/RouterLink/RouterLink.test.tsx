import React from 'react'
import sinon from 'sinon'

import { isErrorLike } from '@sourcegraph/common'
import { renderWithRouter } from '@sourcegraph/shared/src/testing/render-with-router'

import { RouterLink } from './RouterLink'

describe('RouterLink', () => {
    it('renders router link correctly', () => {
        const { asFragment } = renderWithRouter(<RouterLink to="/docs">Link to docs</RouterLink>)
        expect(asFragment()).toMatchSnapshot()
    })

    it('should throw errors when using RouterLink outside of Router', () => {
        const environmentStub = sinon.stub(process.env, 'NODE_ENV').value('development')

        try {
            renderWithRouter(<RouterLink to="/docs">Link to docs</RouterLink>)
        } catch (error) {
            if (isErrorLike(error)) {
                expect(error.message).toBe('Please use the `AnchorLink` component outside of `react-router`')
            } else {
                throw new Error('Unexpected errors')
            }
        }

        environmentStub.restore()
    })
})
