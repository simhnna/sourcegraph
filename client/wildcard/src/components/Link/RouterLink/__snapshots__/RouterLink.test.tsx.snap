// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouterLink renders router link correctly 1`] = `
<DocumentFragment>
  <a
    class="anchorLink"
    href="/docs"
  >
    Link to docs
  </a>
</DocumentFragment>
`;
